### ISSUE TITLE
Notifications : Accessibility: Add Keyboard Navigation To Notification Buttons

### ISSUE THREAD

**Issue by cleidigh** [Microsoft/vscode#32034]
Currently the only keyboard control of notification messages is Escape to clear the message
or messages.
There is an increasing number of notification messages with multiple important button options
and the only way to address them is with the mouse.

Proposal:
- Add Focus to notification message keyboard command
- Add keyboard controller to the notification widget to allow Left/Right/Up/Down button navigation
- Only remove notification message that has been selected and a button activated

**Comment by derekriemer** [comment]
Depending on the way this is done, it might be useful to set focus to the notification text, then allow tabbing to the buttons (Try bringing up a toast with windows then do win+v, then tab around to see what I mean). The arrow key navigation might be more efficient though.

**Comment by cleidigh** [comment]
@derekriemer 
Thanks for the suggestions.  I am Artie working on this and have thought of the following:
- Initially focus on last action button (typically close)
- Provide a key command to display the full text (for the cases where text is truncated)
- I allowed both Tab and arrow keys for navigation.
I think initially focusing on the most probable action might be best visually using the fewest keystrokes.
check in here later for when I do a PR.

can you provide a pointer to the example you reference?

**Comment by derekriemer** [comment]
I can tell you how to trigger one, assuming you are running win10, but this is not the only way, or necessarily the best way, to do these notifications. I think the method you described is perfect. In windows, you can tab from the notification, through the toast's buttons.  To test this, download my ahk script compiled to exe, (it just lets you clear the clipboard). Then, press control+alt+shift+c, then, win+v to get to the notification, and you can navigate through it with tab and the arrow keys (Depending on the type of toast and the buttons). I generally only tab once to the primary dismiss button. 
https://files.derekriemer.com/clipClear.exe
https://files.derekriemer.com/clipClear.ahk

**Comment by cleidigh** [comment]
@derekriemer 
I am using Windows 7, tried the exe not expecting it to work, doesn't seem to I assume it does need Windows10

**Comment by derekriemer** [comment]
yep

**Comment by cleidigh** [comment]
@derekriemer 
Regardless , I think I get what you've described.
Just for clarification, my current decision to not highlight the text(message title ) is mostly due to the fact
that I cannot expect to change the fundamentals of the UI component.

**Comment by derekriemer** [comment]
That's understandable. If you can ensure it reads out automatically, that would be great.

**Comment by cleidigh** [comment]
@derekriemer 
Hmm, to be honest I have been thinking that the notifications are already aria enabled.  I am assuming when you say 'reads out' you are referring to a screen reader.  If you don't mind me asking, I'm inferring perhaps you are vision impaired?  I have ALS but fortunately still can use my voice not hands.  Everything
I do I do with Dragon voice Control.  Mouse commands are very annoying and inefficient.  I have been focusing my contribution efforts to improve keyboard accessibility, hence this current work.
I also want to make sure I don't break anything and ideally improve the accessibility experience all around.
A couple of questions if you don't mind:
- Any straightforward way to set up a screen reader on Windows 7 so I can test?
- What is the current experience with notifications?
- Can you think of anything that needs to be addressed when making these additions ?

Thanks again for your input, I think this is going to be an important addition.

**Comment by derekriemer** [comment]
if you want to use a screen reader to test, you'll want nvda, then go grab dictationbridge (http://dictationbridge.com), then install the commands if you want screen reader control. 

Currently, it works by yielding an alert, I.E. [1]. This is fine, and if you simply focus the buttons, this should be accessible.
@leonardder might have ideas on whether this works in braille. I certainly agree that less keystrokes here is far better, because you don't have to press a key to find the close button. It would be nice however to have the dialog (or alert) labelled by the alert text, I.E. 
<div role="alertdialog" aria-labelledby="notification-alert-id">
<span id="notification-alert-id">My Notification Text Here</span>
    <div id="buttonscontainer">
        <button>Close</button>
        <button>Do some action</button>
    </div>
</div>

**Comment by cleidigh** [comment]
@bpasero 
Based on your comments for #34422 it sounds like I should abandon my solution for this I was about to PR?
I thought the thumbs up on my proposal above meant was relevant...
Is there's something I can contribute for #22388?

**Comment by bpasero** [comment]
@cleidigh I suggest to hold on to doing PRs on the notification widget until we have figured out our strategy there.

**Comment by bpasero** [comment]
https://github.com/Microsoft/vscode/issues/5538
